### DISCUSSION TITLE
Prisma Client - Delete from junction table

### DISCUSSION THREAD

**Discussion by taylorhlabs** [prisma/prisma#4931] [Q&A] [answered]
How do I delete a record in a junction table? When I try the query below, I get an error. (Related issue #4929 with junction table.)
```ts
const result = await prisma.resourcesTags.delete({
    where: { resourceId: resourceId, tagId: tagId }
  });
```
My Model:
```prisma
model Resource {
  id                      Int                       @id @default(autoincrement())
  userId                  Int                       @map("user_id")
  title                   String
  user                    User                      @relation(fields: [userId], references: [id])
  resourceTags            ResourcesTags[]

  @@index([userId], name: "resources_users_id_idx")
  @@map("resources")
}

model ResourcesTags {
  resourceId Int      @map("resource_id")
  tagId      Int      @map("tag_id")
  resource   Resource @relation(fields: [resourceId], references: [id])
  tag        Tag      @relation(fields: [tagId], references: [id])

  @@id([resourceId, tagId])
  @@index([tagId], name: "tag_resource_id")
  @@map("resources_tags_junction")
}

model Tag {
  id           Int                       @id @default(autoincrement())
  userId       Int                       @map("user_id")
  text         String
  deleted      Boolean                   @default(false)
  user         User                      @relation(fields: [userId], references: [id])
  tagResources ResourcesTags[]

  @@index([userId], name: "tags_users_id_idx")
  @@map("tags")
}
```
Error:
```
Invalid `prisma.resourcesTags.delete()` invocation:

{
  where: {
    resourceId: 175,
    ~~~~~~~~~~
    tagId: 4
    ~~~~~
  }
  ~~~~~~~~~~~~~~~~~~
}

Argument where of type ResourcesTagsWhereUniqueInput needs exactly one argument, but you provided resourceId and tagId. Please choose one. Available args: 
type ResourcesTagsWhereUniqueInput {
  resourceId_tagId?: ResourcesTagsResourceIdTagIdCompoundUniqueInput
}
Unknown arg `resourceId` in where.resourceId for type ResourcesTagsWhereUniqueInput. Did you mean `resourceId_tagId`? Available args:
type ResourcesTagsWhereUniqueInput {
  resourceId_tagId?: ResourcesTagsResourceIdTagIdCompoundUniqueInput
}
Unknown arg `tagId` in where.tagId for type ResourcesTagsWhereUniqueInput. Available args:

type ResourcesTagsWhereUniqueInput {
  resourceId_tagId?: ResourcesTagsResourceIdTagIdCompoundUniqueInput
}


    at Document.validate (/Users/userX/Documents/GitHub/next-auth-prisma-mysql/node_modules/@prisma/client/runtime/index.js:76024:19)
    at NewPrismaClient._executeRequest (/Users/userX/Documents/GitHub/next-auth-prisma-mysql/node_modules/@prisma/client/runtime/index.js:77790:17)
    at /Users/userX/Documents/GitHub/next-auth-prisma-mysql/node_modules/@prisma/client/runtime/index.js:77726:52
    at AsyncResource.runInAsyncScope (node:async_hooks:197:9)
    at NewPrismaClient._request (/Users/userX/Documents/GitHub/next-auth-prisma-mysql/node_modules/@prisma/client/runtime/index.js:77726:25)
    at Object.then (/Users/userX/Documents/GitHub/next-auth-prisma-mysql/node_modules/@prisma/client/runtime/index.js:77845:39)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:93:5) {
  clientVersion: '2.14.0'
}
```
**How do I provide "exactly one argument" when I need to provide both IDs to delete the correct record from the junction table?**

**Comment by ryands17** [comment, marked as answer]
Hey @taylorhlabs 👋 

The issue here is explained in the error:

```
type ResourcesTagsWhereUniqueInput {
  resourceId_tagId?: ResourcesTagsResourceIdTagIdCompoundUniqueInput
}
```

You need to provide the arguments in the same way as follows:

```ts
 const result = await prisma.resourcesTags.delete({
    where: {
      resourceId_tagId: {
        resourceId: resourceId,
        tagId: tagId,
      },
    },
  })
```

If you press <kbd>ctrl+space</kbd> inside the where block in VSCode, you will the required suggestion :)

**Reply by taylorhlabs** [reply]
Thanks for the code snippet. This is exactly what I needed. Is there a specific part of the docs that covers this example? I couldn't locate the exact scenario. Thanks again!

**Reply by ryands17** [reply]
I'm not sure if we have this exact use case in the docs but let me check :) 
Also in this cases autocomplete and TS support will always come to the rescue so don't forget to check for hints and suggestions via the keyboard shortcut.

**Reply by todordev** [reply]
Thank you very much for the answer!
However, I have a question. :)

### How to delete from the junction table by one argument?

I would like to do this request but It does not work.

```javascript
await prisma.sentencemldata.delete({
    where: {
        sentenceid: sentenceId
    }
})
```
  
  The error message is:

```
PrismaClientValidationError:
Invalid `prisma.sentencemldata.delete()` invocation:

{
  where: {
    sentenceid: 1
    ~~~~~~~~~~
  }
}

Unknown arg `sentenceid` in where.sentenceid for type sentencemldataWhereUniqueInput. Did you mean `select`? Available args:
type sentencemldataWhereUniqueInput {
  mldataid_sentenceid?: sentencemldataMldataidSentenceidCompoundUniqueInput
```

**Reply by ryands17** [reply]
@ITPrism 👋 

Your query and error don't match, but as per the error I can say that you need to use `deleteMany` for this.

This is a many-to-many relation and so there can be multiple `resourceId` connected to records. So your query would be:

```ts
const result = await prisma.resourcesTags.deleteMany({
    where: {
      resourceId: resourceId
    },
})
```

**Reply by todordev** [reply]
That's it. :)
Thank you very much!

**Reply by kladnik** [reply]
@ryands17 When using `@@unique(fields: [userId, lineId], name: "test")`, I get an error if I do 
```js
prisma.line.delete({
    where: {
        test: {
            lineId: lineId,
            userId: uid,
        },
    },
});
```
Shouldn't this be the similar as with `@@id` since both add a unique constraint?

**Reply by ryands17** [reply]
@kladnik 👋 

This works for me. What issue are you facing and did you run `prisma generate` to get the correct typing?

**Reply by kladnik** [reply]
Oh wow, I really though I ran `prisma generate`, but I'm guessing I didn't. Thanks, it works now. Sorry for confusion!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
